From: Sergey Senozhatsky <senozhatsky@chromium.org>
Subject: zsmalloc: make huge class watermark zs_pool member
Date: Mon, 31 Oct 2022 14:41:03 +0900

We will permit per-pool configuration of pages per-zspage value, which
changes characteristics of the classes and moves around huge class size
watermark.  Thus huge class size needs to be a per-pool variable.

Link: https://lkml.kernel.org/r/20221031054108.541190-5-senozhatsky@chromium.org
Signed-off-by: Sergey Senozhatsky <senozhatsky@chromium.org>
Cc: Alexey Romanov <avromanov@sberdevices.ru>
Cc: Minchan Kim <minchan@kernel.org>
Cc: Nitin Gupta <ngupta@vflare.org>
Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
---

 mm/zsmalloc.c |   10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

--- a/mm/zsmalloc.c~zsmalloc-make-huge-class-watermark-zs_pool-member
+++ a/mm/zsmalloc.c
@@ -177,7 +177,6 @@ static struct dentry *zs_stat_root;
  * (see: fix_fullness_group())
  */
 static const int fullness_threshold_frac = 4;
-static size_t huge_class_size;
 
 struct size_class {
 	struct list_head fullness_list[NR_ZS_FULLNESS];
@@ -231,6 +230,7 @@ struct zs_pool {
 
 	u32 num_size_classes;
 	u32 min_alloc_size;
+	size_t huge_class_size;
 
 	struct zs_pool_stats stats;
 
@@ -1445,7 +1445,7 @@ EXPORT_SYMBOL_GPL(zs_unmap_object);
  */
 size_t zs_huge_class_size(struct zs_pool *pool)
 {
-	return huge_class_size;
+	return pool->huge_class_size;
 }
 EXPORT_SYMBOL_GPL(zs_huge_class_size);
 
@@ -2383,8 +2383,8 @@ struct zs_pool *zs_create_pool(const cha
 		 * endup in the huge class.
 		 */
 		if (pages_per_zspage != 1 && objs_per_zspage != 1 &&
-				!huge_class_size) {
-			huge_class_size = size;
+				!pool->huge_class_size) {
+			pool->huge_class_size = size;
 			/*
 			 * The object uses ZS_HANDLE_SIZE bytes to store the
 			 * handle. We need to subtract it, because zs_malloc()
@@ -2394,7 +2394,7 @@ struct zs_pool *zs_create_pool(const cha
 			 * class because it grows by ZS_HANDLE_SIZE extra bytes
 			 * right before class lookup.
 			 */
-			huge_class_size -= (ZS_HANDLE_SIZE - 1);
+			pool->huge_class_size -= (ZS_HANDLE_SIZE - 1);
 		}
 
 		/*
_
